terraform task-2 done

sean-ias opened the pull request:
Good day! Finally, finished it successfully! : ) 
Important notes:
- Used outputs.tf config file both in the root dir and modules dir, to expose my output from the resources to the root dir module. Or else, I wasn't getting any output after the "terraform apply" command.
- Used kreuzwerker/docker latest version as the provider.

grommir commented on task2_terraform/main.tf:
```diff
@@ -0,0 +1,9 @@
+module "docker_container" {
```
In task we have next requirements:
![image](https://user-images.githubusercontent.com/25118604/235871967-0a822fa9-eab4-4b8e-845a-ffac82d1b38c.png)
This means that the number of containers and images may vary.

sean-ias replied:
ok, I misunderstood it then. I will add this function and make a new commit.

sean-ias commented:
_count = var.container_count
  name  = "${var.container_name}${count.index}"

  ports {
    internal = var.internal_port
    external = var.external_port + count.index
  }_
 - used string concatenation to provide unique names for every container "countainer_name + count.index" where count index is (my_container+0,1,2,3,...)
 - also to provide open unique ports for every container, added count.index to each external port => (8080+0, 8080+1, 8080+2, ...)

sean-ias commented:
- Dramatically changed variables and outputs (from a long minor chunks to one short list of objects = containers)
- in the root dir, moved all content of variables.tf & outputs.tf into main.tf. And got rid of variables.tf and outputs.tf
- changed memory = ... code in the module/docker_container/main.tf 
- new screenshots!

grommir commented:
Looks much better. But what should it look like if we do not want to expose the container to the external network?

sean-ias commented:
> Looks much better. But what should it look like if we do not want to expose the container to the external network?

### **The first solution:**
If we don't want to expose our container to the external network, we can simply **remove the port block** from the resource in our code. This will prevent the container from being accessible from outside the network. **Also, in addition** to that, we can **_set network_mode parameter in the resource block_**, thus, the container will only be accessible within this defined network.

It would look like this:
_resource "docker_container" "container" {
  count = length(var.containers)
  image = var.containers[count.index].image
  name  = var.containers[count.index].name
  **network_mode = "bridge"**
  env = [
    "ENVIRONMENT_VAR = ${var.containers[count.index].environment}"
  ]
  memory = var.containers[count.index].environment == "dev" ? 128 : var.containers[count.index].environment == "stage" ? 256 : var.containers[count.index].environment == "prod" ? 512 : "invalid"
}_

**network_mode=bridge means**, for the containers on bridge network to communicate or be reachable from the outside world, port mapping needs to be configured. So, if we don't configure our port mapping, it means our containers are not accessible externally.

### **The second solution:**
1. remove the port mapping in docker_container resource
2. add a docker_network to create an internal network:
...
_**resource "docker_network" "my_app_net" {
name = "my_app_net"
}**_
...
4. connect the container to this net and add depends_on block to ensure our net is created first:
...
_resource "docker_container" "container" {
count = length(var.containers)
  image = var.containers[count.index].image
  name  = var.containers[count.index].name
  **networks_advanced {
name = docker_network.my_app_net.name
}
depends_on = [docker_network.my_app_net]**
  env = [
    "ENVIRONMENT_VAR = ${var.containers[count.index].environment}"
  ]
  memory = var.containers[count.index].environment == "dev" ? 128 : var.containers[count.index].environment == "stage" ? 256 : var.containers[count.index].environment == "prod" ? 512 : "invalid"
}_
...
5. This will deploy the container on an internal docker net, without any external port mapping, isolating it from external access!

sean-ias commented:
Updates done!:
- Extra network configuration for internal access
- Port configuration for allowing some of our containers to run exposed and some of them unexposed
- Error handling and comments are added
